plugins/inputs/mavlink/mavlink.go

```diff
@@ -0,0 +1,126 @@
+//go:generate ../../../tools/readme_config_includer/generator
+package mavlink
+
+import (
+	_ "embed"
+	"time"
+
+	"github.com/chrisdalke/gomavlib/v3"
+	"github.com/chrisdalke/gomavlib/v3/pkg/dialects/ardupilotmega"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+// Plugin state
+type Mavlink struct {
+	// Config param
+	FcuURL                 string   `toml:"fcu_url"`
+	SystemID               uint8    `toml:"system_id"`
+	MessageFilter          []string `toml:"message_filter"`
+	StreamRequestEnable    bool     `toml:"stream_request_enable"`
+	StreamRequestFrequency int      `toml:"stream_request_frequency"`
+
+	Log telegraf.Logger `toml:"-"`
+
+	// Internal state
+	connection *gomavlib.Node
+	acc        telegraf.Accumulator
+	loading    bool
+	terminated bool
+}
+
+//go:embed sample.conf
+var sampleConfig string
+
+func (*Mavlink) SampleConfig() string {
+	return sampleConfig
+}
+
+// Container for a parsed Mavlink frame
+type metricFrameData struct {
+	name   string
+	tags   map[string]string
+	fields map[string]any
+}
+
+func (s *Mavlink) Start(acc telegraf.Accumulator) error {
+	s.acc = acc
+
+	// Start routine to connect to Mavlink and stream out data async
+	go func() {
+		endpointConfig, err := ParseMavlinkEndpointConfig(s)
+		if err != nil {
+			s.Log.Debugf("%s", err.Error())
+			return
+		}
+
+		// Start MAVLink endpoint
+		s.loading = true
+		s.terminated = false
+		for s.loading {
+			connection, err := gomavlib.NewNode(gomavlib.NodeConf{
+				Endpoints:              endpointConfig,
+				Dialect:                ardupilotmega.Dialect,
+				OutVersion:             gomavlib.V2,
+				OutSystemID:            s.SystemID,
+				StreamRequestEnable:    s.StreamRequestEnable,
+				StreamRequestFrequency: s.StreamRequestFrequency,
+			})
+			if err != nil {
+				s.Log.Debugf("Mavlink failed to connect (%s), will try again in 5s...", err.Error())
+				time.Sleep(5 * time.Second)
+				continue
+			}
```

Comment:
Simply return a `internal.StartupError` and let Telegraf handle the reconnect.

Reply:
Updated so the Mavlink connection happens outside the loop and throws `StartupError`. One caveat I do see about doing it this way is that this causes Telegraf to shutdown if Mavlink client initialization fails, so we could lose the metrics from other plugins if we're rebooting waiting for the USB device which may not turn on at the same time as the companion computer.

That said, this condition actually only happens in serial port mode, where the plugin is looking for a serial device. In TCP or UDP mode the client internally handles reconnection so this won't cause the client to fail.